### ISSUE TITLE
[FEATURE] Create a template from VM, including root and data disks

### ISSUE THREAD

**Issue by yasker** [harvester/harvester#571]
**Is your feature request related to a problem? Please describe.**
Sometimes the user would want to create a template that includes all the states of the VM and use the template to create copies of the VM in the future.

**Describe the solution you'd like**
Support create a VM template with data from an existing VM.

**Describe alternatives you've considered**
VM clone (#569) is another way to solve a similar problem, but it won't work once all the copies are down.

**Additional context**
Depends on #568

**Comment by guangbochen** [comment]
Create a template from VM with includes its root and data disks means users can restore one or multiple new VMs from the template even if the original VM was removed.

With the above condition, I think we can leverage the existing [Create image from volume](https://github.com/harvester/harvester/issues/568) feature to persist `root` disk and `data` disks respectively. 

The only concern that I have is about the disk space availability of those backing image volumes created the VM templates, WDUT cc @shangma @yasker @shuo-wu

**Comment by yasker** [comment]
Yes. We need to create new images from existing VMs to make this feature work. @shuo-wu do we have capacity check for base images?

**Comment by shuo-wu** [comment]
When a Longhorn volume is launched **with a backing image**, the new backing image exported/created from it won't waste space to store zero data **only if the type is qcow2**. But there is one more potential issue besides this: If the upper layer/filesystem writes a file then removed it, the corresponding **sectors were used by the file cannot be reclaimed as free after the file removal**. Then meaningless sector data will be consumed extra space when the volume is exported to a backing image.

e.g.:
1. Upload an empty backing image with apparent size 1Gi. Notice that both the apparent size & actual size of the backing image is 1Gi after upload (I may need to check this part).
```
### before upload, in local machine
~ mkfs.ext4 bi-upload.raw
mke2fs 1.46.1 (9-Feb-2021)
Creating filesystem with 262144 4k blocks and 65536 inodes
Filesystem UUID: c2f3498b-2a71-4eb7-8ee7-02c79d032d70
Superblock backups stored on blocks:
	32768, 98304, 163840, 229376

Allocating group tables: done
Writing inode tables: done
Creating journal (8192 blocks): done
Writing superblocks and filesystem accounting information: done

~ ls -alh bi-upload.raw
-rw-r--r--  1 shuo_wu  staff   1.0G Oct 15 12:36 bi-upload.raw

~ du -h bi-upload.raw
 49M	bi-upload.raw
```
```
### after upload, in the cluster
root@backing-image-manager-c00e-514f:/data/backing-images/bi-upload-abe18172# ls -alh backing
-rw-r--r-- 1 root root 1.0G Oct 15 04:41 backing

root@backing-image-manager-c00e-514f:/data/backing-images/bi-upload-abe18172# du -h backing
1.1G	backing
```
2. Create a volume using the backing image, then make a filesystem for the volume. Now the actual size of the volume is around 100MB.
```
root@shuo-cluster-worker-1:/mnt# mount /dev/longhorn/vol vol/
root@shuo-cluster-worker-1:/mnt# cd vol

root@shuo-cluster-worker-1:/mnt/vol# ls
lost+found

root@shuo-cluster-worker-1:/mnt/vol# dd if=/dev/urandom of=./data1 bs=1M count=100
100+0 records in
100+0 records out
104857600 bytes (105 MB, 100 MiB) copied, 0.682671 s, 154 MB/s

root@shuo-cluster-worker-1:/mnt/vol# sync
root@shuo-cluster-worker-1:/mnt/vol# ls -alh
total 101M
drwxr-xr-x 3 root root 4.0K Oct 15 05:24 .
drwxr-xr-x 5 root root 4.0K Sep 27 04:39 ..
-rw-r--r-- 1 root root 100M Oct 15 05:24 data1
drwx------ 2 root root  16K Oct 15 04:36 lost+found
```
3. Export a raw backing image then a qcow2 backing image from this volume:
```
### raw image. The empty/zero data in the original volume takes space here.
root@backing-image-manager-c00e-f94c:/data/backing-images/bi-export-1-47106e2d# ls -lah backing
-rw-r--r-- 1 root root 1.0G Oct 15 05:28 backing

root@backing-image-manager-c00e-f94c:/data/backing-images/bi-export-1-47106e2d# du -h backing
1.1G	backing

### qcow2 image
root@backing-image-manager-c00e-896f:/data/backing-images/bi-export-2-726d4397# ls -lah backing
-rw-r--r-- 1 root root 102M Oct 15 05:33 backing

root@backing-image-manager-c00e-896f:/data/backing-images/bi-export-2-726d4397# du -h backing
102M	backing
```
4. Write some data in the volume then remove it. Now the actual size of the volume is 600~700MB.
5. Export one more qcow2 backing image from the current volume:
```
root@backing-image-manager-c00e-f94c:/data/backing-images/bi-export-3-d55df0be# ls -lh backing
-rw-r--r-- 1 root root 702M Oct 15 05:52 backing

root@backing-image-manager-c00e-f94c:/data/backing-images/bi-export-3-d55df0be# du -h backing
702M	backing
```

**Comment by shangma** [comment]
### **To Test:**
1. create a vm with 2 or more volume
![image](https://user-images.githubusercontent.com/2396545/138181386-72c158cf-2a67-4617-a57b-7a5d20eaad54.png)

2. Click on "Generate template" and give it a name
![image](https://user-images.githubusercontent.com/2396545/138181420-a3fb3226-9dde-4bfb-89af-fea8378107b2.png)

3. Check if 2 or more images are created based on the same amount of volume from step 1:
![image](https://user-images.githubusercontent.com/2396545/138181475-76f44b15-10ed-4aee-b865-ddfcd47e6111.png)

4. Check if the template is created
![image](https://user-images.githubusercontent.com/2396545/138181507-a97e5137-3280-47e1-949c-e7d9542cdcae.png)

5. Click on the template and check its yaml file to see if the newly created image are connected to this template
![image](https://user-images.githubusercontent.com/2396545/138181546-56a4b62b-2c2d-4398-8e06-34c93b34bafe.png)

**Comment by FrankYang0529** [comment]
One of the use cases for this issue is users can create RKE node by cloning a VM, so I add `area/node-driver` label here. This issue will rely on VM Snapshot feature https://github.com/harvester/harvester/issues/553.

![Screen Shot 2022-05-19 at 5 07 24 PM](https://user-images.githubusercontent.com/4927361/169257020-022534be-7c45-4e08-b287-06411219793e.png)

**Comment by FrankYang0529** [comment]
Hi @yasker, this issue is similar to VM Snapshot (https://github.com/harvester/harvester/issues/553), but you said `VM clone is another way to solve a similar problem, but it won't work once all the copies are down.` I am curious what does `all the copies are down` mean here? 

We use CSI snapshot for VM snapshot, so we rely on mounted volumes when we restore a VM. Do you want to create VMImage for all volumes, so we don't need to rely on origin volumes? Thank you.

**Comment by FrankYang0529** [comment]
Currently, our backup & snapshot have some limitations. For the backup, it needs to set up a backup-target first. For the snapshot, a source volume can't be deleted and must be mounted so that we can restore it.

I think this feature would like to give a third way - don't need to rely on a backup-target and a source volume. We can add [another type `backup-as-backingimage`](https://github.com/harvester/harvester/blob/a4872361145e0d0b99c4afeb6052f344acd48f30/pkg/apis/harvesterhci.io/v1beta1/backup.go#L34-L39). For `backup-as-backingimage`, we will use [BackingImage and set volume in `sourceParameters`](https://longhorn.io/docs/1.3.2/advanced-resources/backing-image/#create-a-backing-image-via-yaml), so we can offer another in-cluster backup method.

cc @guangbochen @bk201 WDYT? Thank you.

**Comment by yasker** [comment]
> Hi @yasker, this issue is similar to VM Snapshot (#553), but you said `VM clone is another way to solve a similar problem, but it won't work once all the copies are down.` I am curious what does `all the copies are down` mean here?
> 
> We use CSI snapshot for VM snapshot, so we rely on mounted volumes when we restore a VM. Do you want to create VMImage for all volumes, so we don't need to rely on origin volumes? Thank you.

Right. If the source volume is not attached, the VM clone won't work or we need to reattach the source volume first.

I was thinking in line with https://github.com/harvester/harvester/issues/571#issuecomment-943158109 and https://github.com/harvester/harvester/issues/571#issuecomment-948086400 . It would be based on #568 to shorten the steps for the user to create golden images and templates with them.

Before:
1. User navigates to a VM's volumes.
2. Create images (e.g. `img-a`, `img-b`) from each volume (`vol-a`, `vol-b`) of the VM.
4. Create a template (without volume data) from the VM.
5. Add images (e.g. `img-a`, `img-b`) to the template, then save the change.
6. Launch the template to create new VMs.

After
1. User click `Create a template with data` (or another name)
2. The template would be generated automatically after setting a few parameters.
3. User launch the template to create new VMs.

I feel it's more of a frontend/UI change than a backend change.

Also, since users can still achieve the same result via multiple API or multiple steps in the UI today, I don't consider this is a high-priority item for now.

**Comment by TachunLin** [comment]
Verified **fixed** on `master-1222750f-head` (04/25) . Close this issue.

### Result
* `vm1-template` is available after we `generate template` from vm1
![image](https://user-images.githubusercontent.com/29251855/234515137-84e5b6a3-b0bf-44f5-9ee9-0229dfdbd138.png)

* Can be export two images and named as the template version of `vm1-template`
![image](https://user-images.githubusercontent.com/29251855/234515690-92f6e14a-1956-4f31-9f19-28de2a247cec.png)

* The newly created image are connected to this template
![image](https://user-images.githubusercontent.com/29251855/234521746-c061751f-cdf0-4449-8cd2-e89f9b8d57b5.png)


* We can write some data into both root disk and disk-2 again, plus execute `sync` after create `vm2` from template



### Test Information
* Test Environment: 2 nodes local kvm machine
* Harvester version: master-1222750f-head

### Verify Steps
https://github.com/harvester/harvester/issues/3545#issuecomment-1450437814
